### PR TITLE
codec: Support read_bytes and write_bytes

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

###  What have you changed?

This PR adds `read_bytes` and `write_bytes` to `BufferReader` and `BufferWriter`. This can make https://github.com/tikv/tikv/pull/5323 simpler.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

Added new unit tests.

### CODE REVIEW

File: components/codec/src/buffer.rs [hunk 1]
```diff
@@ -15,21 +19,48 @@ pub trait BufferReader {
...
     }
 
+    #[inline]
     fn advance(&mut self, count: usize) {
         let mut pos = self.position();
         pos += count as u64;
         self.set_position(pos);
     }
+
+    fn read_bytes(&mut self, count: usize) -> Result<&[u8]> {
```
Comment (sticnarf) [inline components/codec/src/buffer.rs]: I hope you wrote comments dissuading people from wasting time simplifying the function. I was beaten by the borrow checker. :(

Comment (breeswish) [reply components/codec/src/buffer.rs]: `// Do not ever try to waste time on simplifying this function. Here are failed challengers: xxxx`  ? 😄

File: components/codec/src/buffer.rs [hunk 2]
```diff
@@ -15,21 +19,48 @@ pub trait BufferReader {
     ///
     /// This function may panic in some implementors when remaining space
     /// is not large enough to advance.
+    ///
+    /// TODO: We should make the panic behaviour deterministic.
     fn advance(&mut self, count: usize);
+
+    /// Read next several bytes as a slice and advance the position of internal cursor.
```
Comment (sticnarf) [inline components/codec/src/buffer.rs]: The comment looks weird if the implementor is `&[u8]`

**Pull request merged**